src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabUserManagementService.java

```diff
@@ -0,0 +1,271 @@
+        addUserToGroups(userId, instructorExercises, MAINTAINER);
+        addUserToGroups(userId, teachingAssistantExercises, GUEST);
+    }
+
+    @Override
+    public void updateUser(User user, Set<String> removedGroups, Set<String> addedGroups) {
+        if (removedGroups.isEmpty() && addedGroups.isEmpty()) {
+            return;
+        }
+        try {
+            final var gitlabUser = gitlab.getUserApi().getUser(user.getLogin());
+
+            // Add as member to new groups
+            if (!addedGroups.isEmpty()) {
+                final var exercisesWithAddedGroups = programmingExerciseRepository.findAllByInstructorOrTAGroupNameIn(addedGroups);
+                for (final var exercise : exercisesWithAddedGroups) {
+                    final var accessLevel = addedGroups.contains(exercise.getCourse().getInstructorGroupName()) ? MAINTAINER : GUEST;
+                    gitlab.getGroupApi().addMember(exercise.getProjectKey(), gitlabUser.getId(), accessLevel);
+                }
+            }
+
+            // Update/remove old groups
+            if (!removedGroups.isEmpty()) {
+                final var exercisesWithOutdatedGroups = programmingExerciseRepository.findAllByInstructorOrTAGroupNameIn(removedGroups);
+                for (final var exercise : exercisesWithOutdatedGroups) {
+                    // If the the user is still in another group for the exercise (TA -> INSTRUCTOR or INSTRUCTOR -> TA),
+                    // then we have to add him as a member with the new access level
+                    final var course = exercise.getCourse();
+                    if (user.getGroups().contains(course.getInstructorGroupName())) {
+                        gitlab.getGroupApi().updateMember(exercise.getProjectKey(), gitlabUser.getId(), MAINTAINER);
+                    }
+                    else if (user.getGroups().contains(course.getTeachingAssistantGroupName())) {
+                        gitlab.getGroupApi().updateMember(exercise.getProjectKey(), gitlabUser.getId(), GUEST);
+                    }
+                    else {
+                        // If the user is not a member of any relevant group any more, we can remove him from the exercise
+                        gitlab.getGroupApi().removeMember(exercise.getProjectKey(), gitlabUser.getId());
+                    }
+                }
+                gitlab.getUserApi().updateUser(gitlabUser, userService.decryptPasswordByLogin(user.getLogin()).get());
+            }
+        }
+        catch (GitLabApiException e) {
+            throw new GitLabException("Error while trying to update user in GitLab: " + user, e);
+        }
+    }
+
+    @Override
+    public void updateCoursePermissions(Course updatedCourse, String oldInstructorGroup, String oldTeachingAssistantGroup) {
+        if (oldInstructorGroup.equals(updatedCourse.getInstructorGroupName()) && oldTeachingAssistantGroup.equals(updatedCourse.getTeachingAssistantGroupName())) {
```

Comment:
this method and the one below look quite complex to me. Please add more comments so that it is easier to understand what is happening and more importantly why this is happening!

Reply:
I added lots of new comments. Let me know, what you think